Local Authority API

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Adds a JSON endpoint at /api/local-authority.

Although it's a little different because of the disambuguation step and the single output, we try to follow a common RESTful pattern:

/api/local-authority (the index action) exists as the search interface. We don't really want to just return a paged list of local authorities at this time, but it accepts the query parameter (finding the local authority resource by postcode), which allows it to redirect to the show action in the perfect case (a single authority recognised), or return a disambiguation result (a list of addresses with local authority slugs) if more information is needed.

/api/local-authority/slug (the show action) returns the data about the local authority.

## Why

Allows the App to query local authority information to present links to the user's local authority on their home screen. This effectively replicates the information on https://www.gov.uk/find-local-council, but in a JSON form and (since it's intended for machine use), better 

https://trello.com/c/YX1VcZj0/535-create-json-endpoint-version-of-find-local-council-for-app, [Jira issue PNP-6445](https://gov-uk.atlassian.net/browse/PNP-6445)

KludgeKML committed:
refactor: Update AddressPresenter with additional output option

- we're going to want to return a more complex set of data from this
  presenter, with a displayable local authority name. So we add that
  method and refactor/rename the existing options method to
  component_options, which uses the new method and simply maps the
  output to new keys that the components will use.
- addresses_with_authority_data returns an array of the addresses with
  local authority slug and name, suitable for presenting to the app.

KludgeKML committed:
feature: Add a to_h method to the LocalAuthority model that will neatly return the parent if present.

KludgeKML committed:
docs: Add documentation for local authority API

leenagupte commented on app/presenters/address_list_presenter.rb:
```diff
@@ -1,12 +1,25 @@
+      local_authority = LocalAuthority.from_local_custodian_code(address.local_custodian_code)
+      {
+        address: address.address,
+        slug: local_authority.slug,
+        name: local_authority.name,
```
What is the difference between this method and what `options` is returning? Should that method be updated instead or adding a new one?

It feels like the web page and the json response should be using the same data so that they always match.

leenagupte replied:
I think if the `value` field that's returned by `options` is renamed to `slug` and the `name` field added, you wouldn't need this separate method.

As far as I can tell, the only place that `value` is used is an electoral view:
https://github.com/alphagov/frontend/blob/a26af59d248e2c5b04bfe91f853dd8261907b935/app/views/electoral/address_picker.html.erb#L19

leenagupte replied:
Actually that might not be true. Is `value` being used anywhere?

leenagupte replied:
Ok, so it's being used in the multiple_authorities partial. So perhaps instead, `options` should be updated to use the data from `addresses_with_authority_data` and assigns `value` to `slug`. It does feel that these two methods could be dry-ed up.

leenagupte reviewed (commented):
I think this is on the right track. However, I had a few head-scratching moments trying to figure out what the `index` is doing.

As it's always returning JSON, I don't routes need to redirect, I'd imagined it would be case of just returning the data.

I've added some inline comments with questions and suggestions.

deborahchua commented on spec/requests/local_authority_api_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    context "when a postcode is matched to a single authority" do
+      before { configure_locations_api_and_local_authority("SW1A 1AA", %w[westminster], 5990) }
+
+      it "redirects to the a 404 if there is no query parameter" do
```
Is this meant to say "redirects to the local authority"?

deborahchua commented on spec/requests/local_authority_api_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    context "when a postcode is matched to a single authority" do
+      before { configure_locations_api_and_local_authority("SW1A 1AA", %w[westminster], 5990) }
+
+      it "redirects to the a 404 if there is no query parameter" do
```
Is this meant to say "redirects to the local authority"?

deborahchua replied:
We could also add the expectation `expect(response).to have_http_status(:ok)`

deborahchua reviewed (commented):
This looks good! I agree with Leena that more information in the commits would be useful for context (and links to commits, etc). I've add some small inline comments too.

KludgeKML commented on app/presenters/address_list_presenter.rb:
```diff
@@ -1,12 +1,25 @@
     end
   end
+
+  def addresses_with_authority_data
```
Done

KludgeKML commented on app/controllers/api/local_authority_controller.rb:
Link added

KludgeKML commented on app/controllers/api/local_authority_controller.rb:
```diff
@@ -0,0 +1,24 @@
+module Api
+  class LocalAuthorityController < Api::BaseController
+    def index
```
I'm following the RESTful convention for queries, which is to have the query on the index of the resource. But I think if I reorder the PR a bit, it will be more obvious (I should add the show first, then the index query).

KludgeKML committed:
feature: Add local authority controller with show action

- this is genuinely an API, so rather than serving it as a separate format on the
  existing /find-local-council endpoint, we just make it its own thing, which
  allows us to restrict it / update it independently of the other controller. To
  prevent us from inheriting slimmer and other rendering code from the application
  controller, we derive directly from ActionController::Base.

- We will try to stay as close to a standard RESTful interface as we can, given the
  constraints of postcodes/addresses. The show action can be very simple, though,
  using the simplified LocalAuthority model introduced in https://github.com/alphagov/frontend/pull/4705
  and the to_h method added in the previous commit.

- status codes are simple (200 + data if the slug is found, 404 + error message if the
  slug is not found).

KludgeKML committed:
feature: Add index action with query to local authority controller

- Trying to stay as close to a standard RESTful interface as we can, we note that
  the resource is the local authority, and a common pattern for search is to put the
  search on the index of the resource via a query parameter.

- We don't want at the moment to provide a traditional paged index of all local
  authorities, so the index exists just as the query point, and in the golden path
  we can just redirect directly to the show path introduced in the previous commit.

- Since we've simplified the code in the existing find_local_council controller,
  (https://github.com/alphagov/frontend/pull/4705) the index action can be close to
  a direct copy of the web interface's find action.

- the query is a GET - in the web interface it's a POST, but that's largely
  for privacy. The lookup is extremely cacheable at the CDN level - it might
  change less than one time a year - so we might as well take advantage
  of that.

- status codes are 301 (redirect straight to a local authority url to be added
  in a subsequent commit), 200 (a list of addresses to choose from if the postcode
  crosses boundaries), 404 (postcode is structurally valid but not found in the
  database), or 400 (postcode query parameter is missing or not valid).

KludgeKML commented on docs/local-authorities-api.md:
```diff
@@ -0,0 +1,93 @@
+# Local Authorities API
```
I've reordered and added additional documentation, but essentially the query being on the index is a common pattern for RESTful search (which has always been a bit of a vague question in RESTful interfaces). IMO Putting the /postcode/<POSTCODE> suffix would imply that we were returning postcodes.

leenagupte reviewed (approved):
Thanks for adding more detail to the commit messages and rearranging the order of the the commits. That's made it all much easier to follow. 👍

deborahchua reviewed (approved):
LGTM!
